src/python/pants/goal/anonymous_telemetry.py

```diff
@@ -0,0 +1,184 @@
+
+
+_bugout_access_token = "3ae76900-9a68-4a87-a127-7c9f179d7272"
+_bugout_journal_id = "801e9b3c-6b03-40a7-870f-5b25d326da66"
+_telemetry_docs_url = "https://www.pantsbuild.org/docs/anonymous-telemetry"
+_telemetry_docs_referral = f"See {_telemetry_docs_url} for details"
+
+
+class AnonymousTelemetry(Subsystem):
+    options_scope = "anonymous-telemetry"
+    help = (
+        "Options related to sending anonymous stats to the Pants project, to aid development. "
+        "Repos that opt into this telemetry will be able to access to their own stats."
+    )
+
+    @classmethod
+    def register_options(cls, register):
+        register(
+            "--enabled",
+            advanced=True,
+            type=bool,
+            help=(
+                f"Whether to send anonymous telemetry to the Pants project. Telemetry is sent "
+                f"asynchronously, with silent failure, and does not impact build times or "
+                f"outcomes. {_telemetry_docs_referral}."
+            ),
+        )
+        register(
+            "--repo-id",
+            advanced=True,
+            type=str,
+            help=(
+                f"An anonymized ID representing this repo.  For private repos, you likely want the "
+                f"ID to not be derived from, or algorithmically convertible to, anything "
+                f"identifying the repo.  For public repos the ID may be visible in that repo's "
+                f"config file, so anonymity of the repo is not guaranteed (although user anonymity "
+                f"is always guaranteed). {_telemetry_docs_referral}."
+            ),
+        )
+
+    @property
+    def enabled(self) -> bool:
+        return cast(bool, self.options.enabled)
+
+    @property
+    def repo_id(self) -> str | None:
+        return None if self.options.repo_id is None else cast(str, self.options.repo_id)
+
+
+class AnonymousTelemetryCallback(WorkunitsCallback):
+    def __init__(self, anonymous_telemetry: AnonymousTelemetry) -> None:
+        super().__init__()
+        self._anonymous_telemetry = anonymous_telemetry
+
+    @staticmethod
+    def validate_repo_id(repo_id: str) -> bool:
+        is_valid = re.match(r"^[a-zA-Z0-9-_]{30,60}$", repo_id) is not None
+        if not is_valid:
+            logger.error(
+                "The repo_id must be between 30 and 60 characters long, and consist of only "
+                "alphanumeric characters, dashes and underscores."
+            )
+        return is_valid
+
+    def __call__(
+        self,
+        *,
+        started_workunits: tuple[Workunit, ...],
+        completed_workunits: tuple[Workunit, ...],
+        finished: bool,
+        context: StreamingWorkunitContext,
+    ) -> None:
+        if not finished:
+            return
+
+        if self._anonymous_telemetry.options.is_default("enabled"):
+            logger.warning(
+                f"Please either set `enabled = true` in the [anonymous-telemetry] section of "
```

Comment:
I get the motivation here... but people need to contact the pants team in order to acquire repo id, no?

Reply:
No, they can generate it with `uuidgen` for example. It just needs to be consistent. We actively don't want to know how to map a repo_id to an actual repo name.